tianshou/data/batch.py

```diff
@@ -674,6 +675,14 @@ def empty(batch: 'Batch', index: Union[
         """
         return deepcopy(batch).empty_(index)
 
+    def condense(self):
```

Comment:
Why is this useful? Plus I don't find this name very informative.

Reply:
When ``cat`` does not support to concatenate a list of batches with some fields filled by empty Batch() (the default behavior of ``numpy.concatenate``), but we need this feature for multi-agent RL. To be consistent with ``numpy``, we can remove empty fields first, and then ``cat`` them.

The functionality is to remove fields with empty Batch, it is ok if you have another name.

Reply:
Why not to support empty field in `cat` instead ? Using the "key reservation" paradigm (empty Batch are not regular keys).

Reply:
I tried, but that's very annoying. Take a code snippet for example:
```python
    b1 = Batch(a=np.random.rand(3, 4), common=Batch(c=np.random.rand(3, 5)))
    b2 = Batch(b=torch.rand(4, 3), common=Batch(c=np.random.rand(4, 5)))
    test = Batch.cat([b1, b2])
```
For partial key 'a', 'a' is not a key of b2, so it is perfectly ok to fill zero in b2. If we support b1.a is Batch() in ``cat``, what is the output ``test.a`` supposed to be? Furthermore, if both b1.a and b2.a is Batch(), what is the output ``test.a`` supposed to be?

If these are made clear, I'm happy to support ``cat`` with empty Batch().

Reply:
For shared keys, it is also very annoying. Currently, we are passing ``cat`` to ``np.concatenate`` for shared keys. The support for ``cat`` with Batch() would make the code much complicated, essentially we need to duplicate the code of dealing with partial keys.

Reply:
One possible solution would be to re-define shared keys, like ``keys_map = [set(k for k, v in batch.items() if not (isinstance(v, Batch) and v.is_empty())) for batch in batches]``, omitting keys with empty Batch() when calculating shared keys.

Reply:
My opinion is

> If we support `b1.a` is Batch() in `cat`, what is the output `test.a` supposed to be? 

`Batch()`, since `b1.a` is `Batch()` and `b2.a` does not exist.

> Furthermore, if both `b1.a` and `b2.a` is Batch(), what is the output `test.a` supposed to be?

`Batch()`.

But it should define that `len(Batch()) == 0` I guess?

Reply:
> One possible solution would be to re-define shared keys, like keys_map = [set(k for k, v in batch.items() if not (isinstance(v, Batch) and v.is_empty())) for batch in batches], omitting keys with empty Batch() when calculating shared keys.

Exactly.  You just need to compute 3 set of keys: 
- shared keys -> Not empty for all batches
- partial keys -> Empty for at least one of the batches AND not empty for  at least one of the batches
- reserved keys -> Empty for at least one of the batches

Reply:
> If we support b1.a is Batch() in cat, what is the output test.a supposed to be?

`b1.a` is `np.random.rand(3, 4)`, and `b2.a` is `Batch()`, so it is the "partial keys" case.

Reply:
I think the 3 cases are pretty clear, natural, and quite straightforward to implement. You just need to refine the computation of "shared keys" and "partial keys", and add a new 5 lines of code "reserved keys" handling.

Reply:
> > One possible solution would be to re-define shared keys, like keys_map = [set(k for k, v in batch.items() if not (isinstance(v, Batch) and v.is_empty())) for batch in batches], omitting keys with empty Batch() when calculating shared keys.
> 
> Exactly. You just need to compute 3 set of keys:
> 
> * shared keys -> Not empty for all batches
> * partial keys -> Empty for at least one of the batches AND not empty for  at least one of the batches
> * reserved keys -> Empty for at least one of the batches

Okay, reserved keys are treated the same as partial keys.

@Trinkle23897 
> But it should define that len(Batch()) == 0 I guess?

No, we can first filter out empty batches by ``batches = [b for b in batches if len(b.__dict__)]``.

Reply:
> I think the 3 cases are pretty clear, natural, and quite straightforward to implement. You just need to refine the computation of "shared keys" and "partial keys", and add a new 5 lines of code "reserved keys" handling.

What about ``stack``? The same semantic meaning may apply, too.

Reply:
Yes it should.